File: koma-cli/src/main/java/org/koma/cli/KomaCLI.java
package org.koma.cli;

import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.fusesource.jansi.AnsiConsole;
import org.koma.cli.context.DI;
import picocli.CommandLine;

import static org.fusesource.jansi.AnsiConsole.systemInstall;
import static org.fusesource.jansi.AnsiConsole.systemUninstall;

@Slf4j
public class KomaCLI {

  static {
    systemInstall();
  }

  public static void main(String... args) {
    log.atInfo().log("Starting Koma CLI...");
    val code = DI.INSTANCE.getBean(CommandLine.class).execute(args);
    if (code != 9090) {
      systemUninstall();
      System.exit(code);
    }
  }
}

File: koma-cli/src/main/kotlin/org/koma/cli/KomaCLI.kt
package org.koma.cli

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.toml.TomlFactory
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import com.github.ajalt.clikt.core.main
import com.google.common.util.concurrent.AbstractIdleService
import com.google.common.util.concurrent.ServiceManager
import io.github.cdimascio.dotenv.Dotenv
import kotlin.io.path.Path
import org.fusesource.jansi.AnsiConsole.systemInstall
import org.koin.core.annotation.ComponentScan
import org.koin.core.annotation.Module
import org.koin.core.annotation.Qualifier
import org.koin.core.annotation.Single
import org.koin.core.context.startKoin
import org.koin.ksp.generated.*
import org.koma.cli.command.*
import org.koma.compiler.model.KomaLayout
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

@Module
@ComponentScan
class CliModule {

  @Single
  fun executor(): ExecutorService {
    return Executors.newThreadPerTaskExecutor(
      Thread.ofPlatform()
        .name("koma-cli-", 0)
        .factory()
    )
  }

  @Single
  fun serviceManager(services: List<AbstractIdleService>): ServiceManager {
    return ServiceManager(services)
  }

  @Single
  fun dotenv(): Dotenv {
    return Dotenv.configure().ignoreIfMissing().ignoreIfMalformed().systemProperties().load()
  }
}

fun main(args: Array<String>) {
  val container = startKoin {
    printLogger()
    modules(CliModule().module)
  }
  val koin = container.koin

  val serviceManager = koin.get<ServiceManager>()
  serviceManager.startAsync()
  val root = koin.get<RootCommand>()
  systemInstall()

  root.main(args)

  serviceManager.stopAsync()
}
